GitHubActions as a CI provider

TheOneric opened the pull request:
In https://github.com/wangqr/Aegisub/issues/79 ci was migrated to travis-ci.com, however as far as I understand [Travis' announcement](https://blog.travis-ci.com/2020-11-02-travis-ci-new-billing) it seems like Travis-ci.com will not offer any MacOS builds to free user once the initial credits run out.
Maybe this will be sorted out and another migration won't be necessary, but if it's not, here's a mostly working alternative.

Since GHA seemed like the easiest alternatives with the project being already hosted on GH, I tried  to port the current Travis script to GHA. While Ubuntu autoconf and cmake builds work fine, ironically, I can't get the MacOS build to work.
It would be great if anyone more familiar with MacOS could attempt to get the Mac build working.

Coveralls fails with the same errors as it does in the Travis build and is therefore commented out atm.

I'll leave some more details as a code ""review"" below.

TheOneric committed:
ci: add GHA as a CI provider

Travis.com is limiting MacOS builds for free MacOS users,
see: https://blog.travis-ci.com/2020-11-02-travis-ci-new-billing

TheOneric commented on .github/workflows/gha-ci.yml:
```diff
@@ -0,0 +1,103 @@
+              libpulse-dev libuchardet-dev libwxgtk3.0-gtk3-dev luarocks portaudio19-dev \
+              python3-pip python3-setuptools \
+              pkg-config libgl1-mesa-dev libgl-dev libfreetype6-dev libfontconfig-dev \
+              intltool libboost-all-dev
```
CMake and autoconf builds have different config requirements.
- autoconf: configures and builds fine with only the following boost libs:
  `libboost-dev libboost-chrono-dev libboost-filesystem-dev libboost-locale-dev libboost-locale-dev libboost-regex-dev libboost-system-dev libboost-thread-dev`
- CMake: on the other hand requires more libs and otherwise fails during configure (so I switched to `libboost-all-dev`)

Given that the autoconf build also builds and tests fine with the more limited library set, I suspect the CMake script is misconfigured.

wangqr replied:
List of boost libs required by CMake can be found at https://github.com/wangqr/Aegisub/blob/e35c16dc85bb2eb0c18d1c2ad0d42671d58773de/CMakeLists.txt#L509
If more than that is required during configure, maybe the dependenty info in FindBoost.cmake is wrong.

TheOneric commented on .github/workflows/gha-ci.yml:
```diff
@@ -0,0 +1,103 @@
+          sudo luarocks install moonscript > /dev/null
+          sudo luarocks install uuid       > /dev/null
+
+          git submodule --quiet init
+          git submodule --quiet update vendor/googletest
```
The original Travis script did only do this for autoconf builds, while CMake builds used cmake to compile and install a Travis provided googletest in `/usr/src/googletest`. In case of Linux builds I think it should be possible to install those manually, but I don't know about Homebrew.
If we want to use system googletests, does it even make sense to also use an ancient googletest from `vendor`?

wangqr replied:
`/usr/src/googletest` is provided by the `googletest` package from Ubuntu, not Travis. See https://packages.ubuntu.com/bionic/amd64/googletest/filelist . I have no idea why it is provided in source form without binary.

Finding system provided googletest was not implemented in autoconf build, therefore `vendor/googletest` is always used. It may be possible to [specify its path during make](https://github.com/wangqr/Aegisub/blob/e35c16dc85bb2eb0c18d1c2ad0d42671d58773de/tests/Makefile#L3), but I didn't test that, and it needs to be done manually on different OSs / distos.

TheOneric replied:
Restored old behaviour for CMake and Ubuntu-autoconf builds now also use system googletest.

TheOneric commented on .github/workflows/gha-ci.yml:
```diff
@@ -0,0 +1,103 @@
+      - name: configure
+        run: |
+          if (echo "${{ matrix.os }}" | grep -qE '^macos-') ; then
+            export PATH="/usr/local/opt/gettext/bin:/usr/local/opt/icu4c/sbin:/usr/local/opt/icu4c/bin:$PATH";
+            export CPPFLAGS="-I/usr/local/opt/gettext/include -I/usr/local/opt/icu4c/include";
+            export LDFLAGS="-L/usr/local/opt/gettext/lib -L/usr/local/opt/icu4c/lib";
+            export ACLOCAL_PATH="/usr/local/opt/gettext/share/aclocal";
+            export PKG_CONFIG_PATH="/usr/local/opt/icu4c/lib/pkgconfig";
```
**MacOS:**

I suspect this part is what needs tweaking to get the MacOS build working. I've poked a bit at it *(without success obviously)*:

- With the env as is (taken from Travis), it fails during linking, see: https://github.com/TheOneric/Aegisub/runs/1559097260?check_suite_focus=true#step:5:263
- If nothing is set, it fails to find a working boost+icu during configuration
- The only parts neccessary to get boost+icu working *(after which it fails with the exact same error as when everything is set is)*:
  ```
  export CPPFLAGS="-I/usr/local/opt/icu4c/include";
  export PKG_CONFIG_PATH="/usr/local/opt/icu4c/lib/pkgconfig";
  ```

wangqr replied:
Some packages from homebrew are installed as [keg-only](https://docs.brew.sh/FAQ#what-does-keg-only-mean). They are installed without overriding the macOS-provided ones. When installing keg-only package, homebrew will print instructions on how to use these packages by setting the environment vars. These lines are copied directly from homebrew's instructions. So some of them might be unnecessary.

wangqr replied:
The linking failure seems related to LuaJIT/LuaJIT#449

TheOneric replied:
Thanks for the hints. Unlike a proper autotools build, Aegisub's autoconf+custom Makefile combination does not keep `CPPFLAGS` set at configure-time saved for subsequent builds and env vars exported in a GHA step are normally reset after each step.

Env setup is now separated from the other steps using GHA-specific behaviour to alter the env of the following steps.

TheOneric committed:
ci: use system googletest when possible in GHA

Setting GTEST_ROOT=/usr/src/googletest/googletest does not work due to
missing permissions in that location.

TheOneric commented:
MacOS-autoconf, Ubuntu-autoconf and Ubuntu-CMake builds are now working. Coveralls is broken in both Travis and GHA.
Build-logs *(requires to be logged into GitHub to see full log)*: https://github.com/TheOneric/Aegisub/actions/runs/426998740
